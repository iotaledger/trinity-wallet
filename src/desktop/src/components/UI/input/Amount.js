import React from 'react';
import PropTypes from 'prop-types';
import classNames from 'classnames';
import { getCurrencySymbol } from 'libs/currency';
import css from 'components/UI/input/Input.css';

import Chevron from 'images/chevron-down-white.png';

const units = ['$', 'i', 'Ki', 'Mi', 'Gi', 'Ti'];

export default class AddressInput extends React.PureComponent {
    static propTypes = {
        amount: PropTypes.string.isRequired,
        balance: PropTypes.number.isRequired,
        settings: PropTypes.object.isRequired,
        label: PropTypes.string.isRequired,
        labelMax: PropTypes.string.isRequired,
        onChange: PropTypes.func.isRequired,
    };

    state = {
        unit: 'Mi',
    };

    onChange = value => {
        value = value.replace(/,/g, '.');
        const trailingDot = value[value.length - 1] === '.' && value.match(/^\d+(\.\d{0,20})?$/g) ? '.' : '';
        value =
            !value.length || value.match(/^\d+(\.\d{0,20})?$/g) ? value * this.getUnitMultiplier() : this.props.amount;
        this.props.onChange(value + trailingDot);
    };

    getUnitMultiplier() {
        let multiplier = 1;
        switch (this.state.unit) {
            case 'i':
                break;
            case 'Ki':
                multiplier = 1000;
                break;
            case 'Mi':
                multiplier = 1000000;
                break;
            case 'Gi':
                multiplier = 1000000000;
                break;
            case 'Ti':
                multiplier = 1000000000000;
                break;
            case '$':
                multiplier = 1000000 * this.props.settings.conversionRate;
                break;
        }
        return multiplier;
    }

    maxAmount = e => {
        e.preventDefault();
        this.props.onChange(this.props.balance.toString());
    };

    unitChange = e => {
        e.preventDefault();

        const index = units.indexOf(this.state.unit) + 1;

        this.setState({
            unit: index === units.length ? units[0] : units[index],
        });
    };

    render() {
        const { amount, balance, settings, label, labelMax } = this.props;
        const { unit } = this.state;

        return (
            <div className={css.input}>
                <fieldset>
                    <a onClick={this.unitChange}>
                        <span>
                            {unit === '$' ? getCurrencySymbol(settings.currency) : unit}
                            <img src={Chevron} />
                        </span>
                    </a>
                    <input
                        type="text"
                        value={amount / this.getUnitMultiplier() + (amount[amount.length - 1] === '.' ? '.' : '')}
                        onChange={e => this.onChange(e.target.value)}
                    />
                    <small>{label}</small>
                </fieldset>
                <a
                    className={classNames(css.checkbox, parseInt(amount) === balance ? css.on : css.off)}
                    onClick={this.maxAmount}
                >
                    {labelMax}
                </a>
            </div>
        );
    }
}
