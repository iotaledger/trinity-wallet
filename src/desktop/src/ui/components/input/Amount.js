import React from 'react';
import PropTypes from 'prop-types';
import classNames from 'classnames';
import { round } from 'libs/util';
import { getCurrencySymbol } from 'libs/currency';

import Icon from 'ui/components/Icon';
import css from './input.css';

const units = ['i', 'Ki', 'Mi', 'Gi', 'Ti', '$'];

/**
 * Ammount input component
 */
export default class AddressInput extends React.PureComponent {
    static propTypes = {
        /** Current ammount value */
        amount: PropTypes.string.isRequired,
        /** Max available ammount */
        balance: PropTypes.number.isRequired,
        /** Fiat currency settings
         * @property {string} conversionRate - Active currency conversion rate to MIota
         * @property {string} currency - Active currency name
         * @property {string} usdPrice - Current USD price per Miota
         */
        settings: PropTypes.shape({
            conversionRate: PropTypes.number.isRequired,
            currency: PropTypes.string.isRequired,
            usdPrice: PropTypes.number.isRequired,
        }).isRequired,
        /** Ammount input label */
        label: PropTypes.string.isRequired,
        /** Max ammount controller label */
        labelMax: PropTypes.string.isRequired,
        /** Ammount change event function
         * @param {string} value - Current ammount value
         */
        onChange: PropTypes.func.isRequired,
    };

    state = {
        unit: 'Mi',
    };

    onChange = (value) => {
        value = value.replace(/,/g, '.');
        const trailingDot = value[value.length - 1] === '.' && value.match(/^\d+(\.\d{0,20})?$/g) ? '.' : '';
        value =
            !value.length || value.match(/^\d+(\.\d{0,20})?$/g) ? value * this.getUnitMultiplier() : this.props.amount;

        this.props.onChange(value + trailingDot);
    };

    getUnitMultiplier(unit) {
        let multiplier = 1;
        const target = unit || this.state.unit;
        switch (target) {
            case 'i':
                break;
            case 'Ki':
                multiplier = 1000;
                break;
            case 'Mi':
                multiplier = 1000000;
                break;
            case 'Gi':
                multiplier = 1000000000;
                break;
            case 'Ti':
                multiplier = 1000000000000;
                break;
            case '$':
                multiplier = 1000000 / (this.props.settings.usdPrice * this.props.settings.conversionRate);
                break;
        }
        return multiplier;
    }

    maxAmount = () => {
        const { amount, balance } = this.props;

        const total = balance === parseInt(amount) ? '0' : balance.toString();

        this.props.onChange(total);
    };

    unitChange = (unit) => {
        const { amount, settings } = this.props;

        if (unit === '$') {
            const fiat = round(amount * settings.usdPrice / 1000000 * settings.conversionRate * 100) / 100;
            this.props.onChange(round(fiat / settings.usdPrice));
        }

        this.setState({
            unit: unit,
        });
    };

    render() {
        const { amount, balance, settings, label, labelMax } = this.props;
        const { unit } = this.state;

        return (
            <div className={css.input}>
                <fieldset>
                    <a>
                        <strong>
                            {unit === '$' ? getCurrencySymbol(settings.currency) : unit}
                            <Icon icon="chevronDown" size={8} />
                            <ul className={css.dropdown}>
                                {units.map((item) => {
                                    return (
                                        <li
                                            key={item}
                                            className={item === unit ? css.selected : null}
                                            onClick={() => this.setState({ unit: item })}
                                        >
                                            {item === '$' ? getCurrencySymbol(settings.currency) : item}
                                        </li>
                                    );
                                })}
                            </ul>
                        </strong>
                        {amount > 0 && unit !== '$' ? (
                            <p>
                                = {getCurrencySymbol(settings.currency)}{' '}
                                {(
                                    round(amount * settings.usdPrice / 1000000 * settings.conversionRate * 100) / 100
                                ).toFixed(2)}
                            </p>
                        ) : null}
                    </a>
                    <input
                        type="text"
                        value={amount / this.getUnitMultiplier() + (amount[amount.length - 1] === '.' ? '.' : '')}
                        onChange={(e) => this.onChange(e.target.value)}
                    />
                    <small>{label}</small>
                </fieldset>
                {balance > 0 && (
                    <a
                        className={classNames(css.checkbox, parseInt(amount) === balance ? css.on : css.off)}
                        onClick={this.maxAmount}
                    >
                        {labelMax}
                    </a>
                )}
            </div>
        );
    }
}
