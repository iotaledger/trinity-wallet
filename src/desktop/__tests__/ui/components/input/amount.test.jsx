import React from 'react';
import { shallow } from 'enzyme';

import Amount from 'ui/components/input/Amount';

const props = {
    amount: '1500000',
    balance: 4000000,
    settings: {
        conversionRate: 1,
        currency: 'USD',
        usdPrice: 1,
    },
    label: 'Foo',
    labelMax: 'Bar',
    onChange: jest.fn(),
};

describe('Amount input component', () => {
    test('Render the component', () => {
        const wrapper = shallow(<Amount {...props} />);

        expect(wrapper).toMatchSnapshot();
    });

    test('Amount value', () => {
        const wrapper = shallow(<Amount {...props} />);

        expect(wrapper.find('input').props().value).toEqual(1.5);
    });

    test('Amount monetary value', () => {
        const wrapper = shallow(<Amount {...props} />);

        expect(wrapper.find('p').text()).toEqual('= $ 1.50');
    });

    test('Amount change callback', () => {
        const wrapper = shallow(<Amount {...props} />);

        wrapper.find('input').simulate('change', { target: { value: '2' } });
        expect(props.onChange).toHaveBeenLastCalledWith('2000000');
    });

    test('Amount set max callback', () => {
        const wrapper = shallow(<Amount {...props} />);

        wrapper.find('a.checkbox').simulate('click');
        expect(props.onChange).toHaveBeenLastCalledWith('4000000');
    });

    test('Address label', () => {
        const wrapper = shallow(<Amount {...props} />);

        expect(wrapper.find('small').text()).toEqual('Foo');
    });
});
