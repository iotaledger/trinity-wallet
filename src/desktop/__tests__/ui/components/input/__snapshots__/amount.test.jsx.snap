// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Amount input component Render the component 1`] = `
<div
  className="input"
>
  <fieldset>
    <a>
      <strong>
        Mi
        <Icon
          icon="chevronDown"
          size={8}
        />
        <ul
          className="dropdown"
        >
          <li
            className={null}
            key="i"
            onClick={[Function]}
          >
            i
          </li>
          <li
            className={null}
            key="Ki"
            onClick={[Function]}
          >
            Ki
          </li>
          <li
            className="selected"
            key="Mi"
            onClick={[Function]}
          >
            Mi
          </li>
          <li
            className={null}
            key="Gi"
            onClick={[Function]}
          >
            Gi
          </li>
          <li
            className={null}
            key="Ti"
            onClick={[Function]}
          >
            Ti
          </li>
          <li
            className={null}
            key="$"
            onClick={[Function]}
          >
            $
          </li>
        </ul>
      </strong>
      <p>
        =
         
        $ 1.50
      </p>
    </a>
    <input
      onChange={[Function]}
      type="text"
      value={1.5}
    />
    <small>
      Foo
    </small>
  </fieldset>
  <a
    className="checkbox off"
    onClick={[Function]}
  >
    Bar
  </a>
</div>
`;
